main/core/src/eval/Evaluator.scala

```diff
@@ -598,221 +547,89 @@ case class Evaluator(
+        //      println("INITIALIZING futures")
+        val futures = terminals.map { k =>
+          val deps = interGroupDeps((k, sortedGroups.lookupKey(k))).map(_._1)
+          //          println("REGISTERING " + label(k) + " <- " + deps)
+          Future.sequence(deps.map(taskFutures(_)))
+            .map { upstreamValues =>
+              //              println("START " + label(k) + " " + deps.size)
+              //              println("THREAD " + Thread.currentThread().getName())
+              //              println("DEPS " + deps)
+              val startTime = System.currentTimeMillis()
+
+              val res = blocking(evaluateGroupCached(k, sortedGroups.lookupKey(k), results, "lol", _ => None, testReporter, logger))
+              val endTime = System.currentTimeMillis()
+              Evaluator.this.synchronized {
 
                 timeLog.timeTrace(
-                  task = term,
+                  task = label(k) + " " + System.identityHashCode(k),
                   cat = "job",
                   startTime = startTime,
                   endTime = endTime,
                   thread = Thread.currentThread().getName(),
                   cached = res.cached
                 )
-                timeLog.debug(s"END   [${counterMsg}] ${term} (${
-                  if (res.newResults.exists(task => !task._2.isInstanceOf[Success[_]])) "failed, " else ""
-                }${
-                  if (res.cached) "cached, " else ""
-                }${
-                  (endTime - startTime).toInt
-                })")
-
-                FutureResult(curWork, (endTime - startTime).toInt, res)
+                for ((k, v) <- res.newResults) results(k) = v
               }
+              //              println("DONE " + label(k) + " " + deps.size)
+              //              println("RES " + Evaluator.this.synchronized((results.keys)))
+              promises(k).success(123)
+              res
             }
 
-            evalLog.debug(s"New future: ${workerFut} [${index + 1}/${newInProgress.size}] for task: ${term}")
-            state.scheduledFutures += workerFut -> curWork
         }
-        timeLog.timeTrace("Schedule", "schedule", scheduleStart, System.currentTimeMillis(), Thread.currentThread().getName(), false)
-      }
-    }
 
-    def printException(e: Throwable): String = {
-      val baos = new ByteArrayOutputStream()
-      val os = new PrintStream(baos)
-      try {
-        e.printStackTrace(os)
-        baos.toString()
-      } finally {
-        os.close()
-        baos.close()
+        //      println("AWAITING futures")
+        val finished = futures.map(Await.result(_, duration.Duration.Inf))
+
+        val failing = new util.MultiBiMap.Mutable[Either[Task[_], Labelled[_]], Result.Failing[_]]
+        for ((k, vs) <- sortedGroups.items()) {
+          failing.addAll(
+            k,
+            vs.items.flatMap(results.get).collect { case f: Result.Failing[_] => f.map(_._1) }
```

Comment:
This is mostly copy-pasted from the sequential evaluator; I can move them into a helper function